### PR TITLE
Add first function to estimate inversion height (and update profile plots with minor changes)

### PR DESCRIPTION
**Pull request by JuleRadtke**
Issue #11

### CODE REVIEW

**Comment by JuleRadtke** [comment]
@annalea-albright I hope I got your function right with the z_min, z_max?

File: eurec4a_environment/plot/profile.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    except ValueError:
+        org_dim = ds[time].dims[0]
+        ds = ds.swap_dims({org_dim: time})
+
+    # Plot
+    fig, ax = plt.subplots(figsize=(20, 10))
+
+    var = ds[variable]
+
+    var.plot.contourf(
```
Comment (Geet-George) [inline eurec4a_environment/plot/profile.py:51]: For cases with irregular time intervals or "discontinuities", using pcolormesh instead of contourf might be better (e.g. for airborne platforms). Maybe we can provide an option for the user to decide between pcolormesh or contourf. The option can be simply interpolate_between_profiles=True or False, so that they don't have to bother with the function name.

File: eurec4a_environment/plot/profile.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    plt.close()
+    return fig
+    # fig.savefig("ProfilePlot2D_{}.pdf".format(variable), bbox_inches="tight")
+
+
+def plot_profile_1D(
+    ds_plot,
+    variables=["rh", "q"],
+    axis_labels=["relative humidity (%)", "specific humidity (g/kg)"],
+    height_labels=None,
```
Comment (Geet-George) [inline eurec4a_environment/plot/profile.py:88]: Maybe **kwargs here too? In case user wants to change some plot properties?

File: eurec4a_environment/plot/profile.py (hunk 1 above)
Comment (JuleRadtke) [reply eurec4a_environment/plot/profile.py:51]: Okay!

File: eurec4a_environment/plot/profile.py (hunk 2 above)
Comment (JuleRadtke) [reply eurec4a_environment/plot/profile.py:88]: True, jip I add those

**Comment by JuleRadtke** [comment]
But somehow I didn't wanted to commit for the plot files in this branch, had a spearate branch for this, seems I did something wrong there...

**Comment by Geet-George** [comment]
I am not sure I understand what you are saying... Any commits to the branch through which you made this pull request should be updated here. Maybe you made a commit to a different branch (e.g. master)?

**Comment by JuleRadtke** [comment]
I openend this branch inversion_heights new just for the inversion function, commited this and made a pull request for this branch... The profiles I head in a separate branch, which is now already merged I think.  - at least thats what I thought I do, but mmh it seems not

**Comment by JuleRadtke** [comment]
But well what I ask myself is if I should now go back to the profile branch (lathough it is already merged) and update the changes there or stay in this branch?

**Comment by JuleRadtke** [comment]
Oh probably I have to checkout -b a new branch always from the master branch?

File: eurec4a_environment/plot/profile.py (hunk 1 above)
Comment (leifdenby) [reply eurec4a_environment/plot/profile.py]: Sounds good to me too ⭐

**Comment by leifdenby** [comment]
If you don't want the `eurec4a_environment/plot/profile.py` to be included in this pull request you can just make a commit which deletes that file

```
git rm eurec4a_environment/plot/profile.py
git commit ...
git push ...
```

If you want the file later you just need to find a commit where it still existed:

```
git checkout <commit-id> eurec4a_environment/plot/profile.py
# for example:
git checkout 202d5a7 eurec4a_environment/plot/profile.py
```

File: eurec4a_environment/plot/profile.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    except ValueError:
+        org_dim = ds[time].dims[0]
+        ds = ds.swap_dims({org_dim: time})
+
+    # Plot
+    fig, ax = plt.subplots(figsize=(20, 10))
+
+    var = ds[variable]
+
+    if interpolate_between_profiles:
```
Comment (leifdenby) [inline eurec4a_environment/plot/profile.py:52]: I would just make a `plot_func` variable say :)

```python
if interpolate_between_profiles:
    plot_func = var.plot.contourf
else:
    plot_func = var.plot.pcolormesh
```

Since you're passing in the same arguments to both

File: eurec4a_environment/variables/boundary_layer/inversion_height.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+            ds_lowertroposhere[rh]
+            .rolling(alt=window, min_periods=window, center=True)
+            .mean(skipna=True)
+        )
+    else:
+        RH = ds_lowertroposhere[rh]
+
+    RHg = RH.differentiate(coord=altitude)
+    ix = RHg.argmax(dim=altitude, skipna=True)
+    da_z = RHg.alt[ix]
```
Comment (leifdenby) [inline eurec4a_environment/variables/boundary_layer/inversion_height.py:21]: this would be better as
```python
da_z = RHg.isel({ altitude: ix }).alt
```

to make it explicit that you're indexing on the height coordinate (just in case someone passes in a dataset where height isn't the first index

File: eurec4a_environment/variables/boundary_layer/inversion_height.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+def find_inversion_height_grad_RH(
+    ds, altitude="alt", rh="rh", smooth=False, z_min=1500, z_max=4000.0
+):
+
+    """
+        Returns inversion height defined as the maximum in the vertical gradient of RH
+        """
+    ds_lowertroposhere = ds.sel({altitude: slice(z_min, z_max)})
+    window = 10  # window for convolution
```
Comment (leifdenby) [inline eurec4a_environment/variables/boundary_layer/inversion_height.py:9]: I would make `window` a kwarg for the function. Maybe you could remove `smooth` and have `smoothing_win_size=None` and then only apply smoothing if the user passes in a number? What do you think?

File: tests/test_boundary_layer.py [hunk 6]
```diff
@@ -21,3 +21,11 @@ def test_mixed_layer_height_RHmax(ds_isentropic_test_profiles):
         ds=ds, altitude="alt", rh="rh"
     )
     assert np.allclose(da_rh_peak, z0)
+
+    import numpy as np
+
+def test_inversion_height_gradient_RH(ds_isentropic_test_profiles):
```
Comment (leifdenby) [inline tests/test_boundary_layer.py:27]: Awesome ! 🌟

Only there's a `import numpy as np` flying around there. Are you using that? :)

File: eurec4a_environment/plot/profile.py (hunk 3 above)
Comment (JuleRadtke) [reply eurec4a_environment/plot/profile.py:52]: True!

File: eurec4a_environment/variables/boundary_layer/inversion_height.py (hunk 4 above)
Comment (JuleRadtke) [reply eurec4a_environment/variables/boundary_layer/inversion_height.py:21]: True thanks!

File: eurec4a_environment/variables/boundary_layer/inversion_height.py (hunk 5 above)
Comment (JuleRadtke) [reply eurec4a_environment/variables/boundary_layer/inversion_height.py:9]: Nice idea yes!

File: tests/test_boundary_layer.py (hunk 6 above)
Comment (JuleRadtke) [reply tests/test_boundary_layer.py:27]: Mmh ja that shouldn't be there...

**Comment by JuleRadtke** [comment]
I deleted the notebook, the other files are all updated with the proposed changes and so the pull request is ready to be merged I think.

**Comment by JuleRadtke** [comment]
And now I did a mess because I tried to resolve the conflict and didn't pulled in the changes to my local repo before adding new changes...

**Comment by leifdenby** [comment]
I just looks like the test you written isn't quite going through because you're testing for `> 1500` but the height you've calculated is exactly `1500`: https://github.com/eurec4a/eurec4a-environment/runs/932187259#step:8:50

**Comment by Geet-George** [comment]
I think the isentropic ds is not a good choice for the test here since there is no gradient in RH, and so, essentially the test would give the first value of z, i.e. 1500 m, which is what @leifdenby 's point is that the function allows for 1500, but the test does not. 

More general question... For inversion heights, why do we have a limit on z (unless it is really low, like 50-100 m)?

**Comment by leifdenby** [comment]
> More general question... For inversion heights, why do we have a limit on z (unless it is really low, like 50-100 m)?

Yes, it would probably be best to just make a profile which has an inversion height at a pre-determined level. You could take the isentropic atmosphere and make it get warmer above a specific height

```python
ds = ds_isentropic.copy()
z_INV = 2000.
dzdT_INV = 0.1*e-3 # I'm not sure if 0.1K/km is the right lapse rate, but it does something...
ds.T = ds.T.where(ds.alt < z_INV, other=(ds.alt - z_INV)*dzdT_INV)
```

**Comment by Geet-George** [comment]
This is a good idea but the function tests for changes in the gradient of RH which is constant throughout the profile, and not temperature. 

I also think since the term inversion is in some way self-defining, we should stick to methods for inversions that depend on T or theta or other forms of temperature, and avoid moisture. This function here of `gradient_RH` can be used to know where the cloud layer top or hydrolapse is, and an inversion need not always be present at these levels. 

Also, why remove `peak_RHlinearise`? Isn't that for the mixed layer?

P.S. ~6 K/km is the dry adiabatic lapse rate.

**Comment by annalea-albright** [comment]
Re Geet's points
>"More general question... For inversion heights, why do we have a limit on z (unless it is really low, like 50-100 m)?"  &  "I also think since the term inversion is in some way self-defining, we should stick to methods for inversions that depend on T or theta or other forms of temperature, and avoid moisture. This function here of gradient_RH can be used to know where the cloud layer top or hydrolapse is, and an inversion need not always be present at these levels."

1. I may have had an old flag to limit the search space in there, but we should remove that flag and just see from the test if the values are physical. From the static stability metric, it seems like the mean inversion base height is ~2km, but had one-sigma of 800m, I think, so there could certainly be some lower values that you wouldn't want to artificially constrain 👍 

2. I've seen inversion base be defined from the max dRH/dz definition, but you are right that temperature seems more directly relevant. So I think the most relevant definitions are 1) static stability and hydrolapse-based (i.e. when static stability > 0.1 K/hPa and there is dRH/dz < 0, following Bony and Stevens 2018) 2) dT/dz > 0 and dRH/dz < 0 (fulfill both conditions, following Grindinger 1992, Can et al 2007) and 3) max dT/dz or max dtheta/dz (though this might be noisier?). Having dRH/dz is useful, but you're right that it is probably not sufficient to say that there is an inversion present at these levels

**Comment by Geet-George** [comment]
**Regarding definition**: Agreed. That's a good point. RH is a great constraint on getting inversion levels (especially strong hydrolapses), but I think that temperature should somehow fit in to the estimation (via e.g. static stability)... 

**Regarding bounds**: Based on max N<sup>2</sup>, at least in terms of circle means, most cases have an inversion at ~2 km or below 2.5 km (~40 out of 71 HALO circles; see plot). I realise that this is a different estimate, but it is essentially based on static stability (with some impact from moisture), and there are still ~10 circles with no inversions up to 4 km with this estimate. For these circles, the first inversion is close to 5-5.5 km. Maybe it's not the best way to estimate inversion heights, in which case my whole point stands invalid. I will leave it upto you @annalea-albright 

Suggestion : We could set a bound of 4000 m and term these levels as inversions within the free troposphere, the definition of which would be another discussion in itself.

**Comment by annalea-albright** [comment]
This is great @Geet-George ! Having an inversion base definition that incorporates both temperature and humidity seems like a good framework to work with.

Max N2 seems like a very good and general definition to include because it includes influences from static stability and moisture, as you say!  When you have a chance, maybe you can push your function to broaden our base of definitions for inter-comparison. I find inversion base heights hard to conceptualize because there can be multiple inversions of varying strengths, so I have a hard time understanding if we want to compare each day's strongest inversion, or the first inversion, or inversions occurring roughly in the same height range (i.e. 1-4km, 4-6km?) ... but maybe I just don't have a good picture in mind, and it is actually simpler than I am imagining. Or it is more complicated, and the isotopes and other measurements might help us make sense of this layering of air masses above the cloud layer :)

**Comment by JuleRadtke** [comment]
Great discussion on inversion heights! @annalea-albright and @Geet-George you are more the experts here! So probably I just copy again the code of grad_rh into here as a start off for other, more elaborated functions and close this pull request? (The small changes in the plotting functions I would incorporate in a separate pull request...)

**Comment by JuleRadtke** [comment]
import xarray as xr

def find_inversion_height_grad_RH(
    ds, altitude="alt", rh="rh", smoothing_win_size=None, z_min=1500, z_max=4000.0
):

    """
        Returns inversion height defined as the maximum in the vertical gradient of RH
        """
    ds_lowertroposhere = ds.sel({altitude: slice(z_min, z_max)})
    if smoothing_win_size:
        RH = (
            ds_lowertroposhere[rh]
            .rolling(alt=smoothing_win_size, min_periods=smoothing_win_size, center=True)
            .mean(skipna=True)
        )
    else:
        RH = ds_lowertroposhere[rh]

    RHg = RH.differentiate(coord=altitude)
    ix = RHg.argmax(dim=altitude, skipna=True)
    da_z = RHg.isel({ altitude: ix }).alt
    da_z.attrs["long_name"] = "inversion layer height (from RH gradient)"
    da_z.attrs["units"] = "m"

    return da_z

**Comment by leifdenby** [comment]
> This is a good idea but the function tests for changes in the gradient of RH which is constant throughout the profile, and not temperature.

Yes, I missed that. For the test @JuleRadtke it would be the same idea as before, but simply make the relative humidity high below a certain altitude and the drop to say zero, that way you can check if your function finds the height that has large gradient. That's the intention of this function, right @annalea-albright?

> Also, why remove `peak_RHlinearise`? Isn't that for the mixed layer?

This pull-request should only have included the new function for calculating inversion heights. @JuleRadtke it looks like you've made a commit that modifies `eurec4a_environment/variables/boundary_layer/mixed_layer_height.py ` (and also removes `CHANGELOG.md`). I think you need to pull in the `master` from the `upstream` branch to this branch and push it to github.

```bash
git checkout inversion_heights
git pull upstream master
git push origin inversion_heights
```

That will ensure those files aren't changed.



> **Regarding definition**: Agreed. That's a good point. RH is a great constraint on getting inversion levels (especially strong hydrolapses), but I think that temperature should somehow fit in to the estimation (via e.g. static stability)...

There's a lot of great discussion here about how to define the inversion height. I suggest we move that to a different issue though so that we can discuss that separately. @annalea-albright had implemented two methods in her original code (https://github.com/eurec4a/eurec4a-environment/issues/11). We might want to do more definitions or change them, but let's discuss that here: https://github.com/eurec4a/eurec4a-environment/issues/27

So in summary @JuleRadtke:

1. I think you could write a test where you create a profile with a large negative gradient in relative humidity, does that sound ok?
2. Make sure you've pulled in from `upstream` `master` and push again

**Comment by JuleRadtke** [comment]
> This pull-request should only have included the new function for calculating inversion heights. @JuleRadtke it looks like you've made a commit that modifies `eurec4a_environment/variables/boundary_layer/mixed_layer_height.py ` (and also removes `CHANGELOG.md`). I think you need to pull in the `master` from the `upstream` branch to this branch and push it to github.
> 
> ```shell
> git checkout inversion_heights
> git pull upstream master
> git push origin inversion_heights
> ```
> 
> That will ensure those files aren't changed.

Yes that is what I meant with mess. Okay thanks I try it.

**Comment by JuleRadtke** [comment]
Mmh I get only "already up to date". Do I really have to pull from upstream to master or to inversion_heights?

**Comment by JuleRadtke** [comment]
>     1. I think you could write a test where you create a profile with a large negative gradient in relative humidity, does that sound ok?

And sure I can do that!

File: setup.py [hunk 7]
```diff
@@ -2,7 +2,7 @@
 from setuptools import setup, find_packages
 import versioneer
 
-INSTALL_REQUIRES = ["xarray", "netcdf4", "typhon", "tqdm", "statsmodels"]
+INSTALL_REQUIRES = ["xarray", "netcdf4", "typhon", "tqdm"]
```
Comment (leifdenby) [inline setup.py:5]: I think we need `statsmodels` too, it's just that we don't have a test for @annalea-albright's linear_RH mixed-layer height yet, so this wouldn't get picked up. But I'll help fix this when we talk about git stuff in a moment

File: tests/test_boundary_layer.py [hunk 8]
```diff
@@ -23,10 +22,11 @@ def test_mixed_layer_height_RHmax(ds_isentropic_test_profiles):
...
-    da_lin = boundary_layer.mixed_layer_height.calc_peakRH_linearize(ds)
-    assert len(da_lin) == 10
-    assert np.all(da_lin < 1500.)
+def test_inversion_height_gradient_RH(ds_isentropic_test_profiles):
+    ds = ds_isentropic_test_profiles.copy()
+    z_INV = 2000.
+    ds['RH'] = ds.RH.where(ds.alt < z_INV, other=0.5)
+    print (ds.RH)
+    da_inv = boundary_layer.inversion_height.find_inversion_height_grad_RH(ds=ds, rh = 'RH')
+    print (da_inv)
```
Comment (leifdenby) [inline tests/test_boundary_layer.py:31]: Love the simplicity of this test!  😄  Did you meant to leave the `print` statements in though?

Comment (JuleRadtke) [reply tests/test_boundary_layer.py]: Ups no

**Comment by leifdenby** [comment]
```bash
git checkout 810907d338ba75e51fd4812cb2767b866c35519c eurec4a_environment/variables/boundary_layer/mixed_layer_height.py
```

**Comment by leifdenby** [comment]
Well done ! 🌟 🌟 🌟
